aea/test_tools/test_skill.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""This module contains test case classes based on pytest for AEA skill testing."""
+
+import asyncio
+from pathlib import Path
+from queue import Queue
+from types import SimpleNamespace
+from typing import Optional, Union, cast
+
+from aea.context.base import AgentContext
+from aea.identity.base import Identity
+from aea.multiplexer import AsyncMultiplexer, Multiplexer, OutBox
+from aea.protocols.base import Message
+from aea.skills.base import Skill
+from aea.skills.tasks import TaskManager
+
+
+class BaseSkillTestCase:
+    """A class to test a skill."""
+
+    path_to_skill: Union[Path, str] = Path(".")
+    _skill: Skill
+    _multiplexer: AsyncMultiplexer
+    _outbox: OutBox
+
+    @property
+    def skill(self) -> Skill:
+        """Get the skill."""
+        try:
+            value = self._skill
+        except AttributeError:
+            raise ValueError("Ensure skill is set during setup_class.")
+        return value
+
+    def get_quantity_in_outbox(self) -> int:
+        """Get the quantity of envelopes in the outbox."""
+        return self._outbox._multiplexer.out_queue.qsize()
+
+    def get_message_from_outbox(self) -> Optional[Message]:
+        """Get message from outbox."""
+        if self._outbox.empty():
+            return None
+        envelope = self._multiplexer.out_queue.get_nowait()
+        return envelope.message
+
+    @classmethod
+    def setup(cls) -> None:
```

Comment:
As we talked about before, a test might change the state of the loaded skill and mess up the execution of other tests (e.g. `test_search_behaviour_setup_b` below changes `strategy._is_ledger_tx`).

The code block here in `setup` does not get called before every test function does it? 
If no, perhaps this code block can be moved into a autouse pytest fixture?

Reply:
> The code block here in setup does not get called before every test function does it?
Correct. It gets called only once for all the tests in the test class.

>  If no, perhaps this code block can be moved into a autouse pytest fixture?
I think the burden to avoid side-effects, or to restore the original state, can be left to the developer. It is more in line with `pytest` test classes behaviour.

Reply:
My bad, I was under the impression it would be reset on each method call. Of course it needs to be a clean skill state for each test, that's the whole point..will fix. I disagree that the burden should be on the developer. The skill has too many side-effects. It will render the whole tool largely pointless...

Reply:
My bad; `setup` seems to be called for each method; although I couldn't find confirmation in the relevant documentation:
https://docs.pytest.org/en/2.8.7/xunit_setup.html#class-level-setup-teardown